nexusclient/client.go

```diff
@@ -0,0 +1,725 @@
+package nexusclient
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log/slog"
+	"net/http"
+	"net/url"
+	"os"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/nexus-rpc/sdk-go/nexusapi"
+)
+
+// Options for creating a Client.
+type Options struct {
+	// Base URL of the service.
+	// Optional. If not provided, this client can only be used to deliver operation completions.
+	ServiceBaseURL string
+	// An HTTP Client to use for making HTTP requests.
+	// Defaults to [http.DefaultClient].
+	HTTPClient *http.Client
+	// Max duration to wait for a single get result request.
+	// Enforced if context deadline for the request is unset or greater than this value.
+	//
+	// Defaults to one minute.
+	GetResultMaxRequestTimeout time.Duration
+	// A stuctured logging handler.
+	// Defaults to logging with text format to stderr at info level.
+	LogHandler slog.Handler
+	// Optional marshaler for marshaling objects to JSON.
+	// Defaults to output with indentation.
+	Marshaler nexusapi.Marshaler
+}
+
+// A Client is used to start an operation, get an [OperationHandle] to an existing operation, and deliver operation
+// completions.
+type Client struct {
+	// The options this client was created with after applying defaults.
+	Options        Options
+	serviceBaseURL *url.URL
+	logger         slog.Logger
+}
+
+var (
+	// Package version.
+	Version = "dev" // TODO: Actual version to be set by goreleaser
+	// User agent used to make HTTP requests.
+	UserAgent       = fmt.Sprintf("Nexus-go-sdk/%s", Version)
+	headerUserAgent = "User-Agent"
+	// Error indicating an empty ServiceBaseURL option was used to create a client when making a Nexus service request.
+	ErrEmptyServiceBaseURL = errors.New("empty serviceBaseURL")
+	// Error indicating a non HTTP URL was used to create a [Client].
+	ErrInvalidURLScheme = errors.New("invalid URL scheme")
+
+	// Asynchronous method was used on an [OperationHandle] representing a synchronous operation.
+	ErrHandleForSyncOperation = errors.New("handle represents a synchronous operation")
+
+	errOperationStillRunning = errors.New("operation still running")
+)
+
+// Error that indicates a client encountered something unexpected in the server's response.
+type UnexpectedResponseError struct {
+	// Error message.
+	Message string
+	// The HTTP response. The response body will have already been read into ResponseBody and closed.
+	Response *http.Response
+	// Body extracted from the HTTP response.
+	ResponseBody []byte
+}
+
+// Error implements the error interface.
+func (e *UnexpectedResponseError) Error() string {
+	if nexusapi.IsContentTypeJSON(e.Response.Header) {
+		var failure nexusapi.Failure
+		if err := json.Unmarshal(e.ResponseBody, &failure); err == nil && failure.Message != "" {
+			return fmt.Sprintf("%s: %s", e.Message, failure.Message)
+		}
+
+	}
+	return e.Message
+}
+
+func (c *Client) newUnexpectedResponseError(message string, response *http.Response) error {
```

Comment:
Just because it has a logger for the error and its the only thing constructing it.

I could have had a static method that returned an error and wrapped that in a method with a log statement but that would be the only place where it's consumed.

Reply:
Ah, didn't see the logger use. I think you should consider bubbling that error out. And really, constructing an error shouldn't read an HTTP response, it should be given content to work with. You want to read HTTP responses in a clear place by where you make the HTTP request. Also, logs by where it occurs are helpful because most Go logging logs the filename and line number of the logs which doesn't help here. No need to over-modularize.